Add 2025 review post (#19175)
pypa/warehouse: docs/blog/posts/2025-12-31-pypi-2025-in-review.md

---
title: "PyPI in 2025: A Year in Review"
description: "A look back at the major changes to PyPI in 2025."
authors:
  - di
date: 2025-12-31
tags:
- new features
- security
- organizations
---

As 2025 comes to a close, it's time to look back at another busy year for the Python Package Index. This year, we've focused on delivering critical security enhancements, rolling out powerful new features for organizations, improving the overall user experience for the millions of developers who rely on PyPI every day, and responding to a number of security incidents with transparency.

<!-- more -->

But first, let's look at some numbers that illustrate the scale of PyPI in 2025:

> *   **[NUMBER]** new files published

Author: ```suggestion
*   More than **3.9 million** new files published
```

> *   **[NUMBER]** new projects created

Author: ```
SELECT count(*)
FROM projects
WHERE created >= '2025-01-01 00:00:00'
AND created < '2026-01-01 00:00:00';
```

Author: ```suggestion
*   More than **130,000** new projects created
```

*   **[NUMBER]** petabytes of data transferred

> *   **[NUMBER]** billions of requests served

Author: Ack, will do (I don't think it's possible to combine stats about services)

These numbers are a testament to the continued growth and vibrancy of the Python community. As always, a huge thanks to our [sponsors](https://pypi.org/sponsors/) who make this kind of scale possible.

Let's dive into some of the key improvements we've made to PyPI this year.

## Security First, Security Always

Security is our top priority, and in 2025 we've shipped a number of features to make PyPI more secure than ever before.

### Enhanced Two-Factor Authentication (2FA) for Phishing Resistance

We've made significant improvements to our 2FA implementation, starting with **[email verification for TOTP-based logins](/posts/2025-11-14-login-verification.md)**. This adds an extra layer of security to your account by requiring you to confirm your login from a trusted device, when using a phishable 2FA method like TOTP.

Since rolling out these changes, we've seen:

> *   **[PERCENTAGE]%** of active users with non-phishable 2FA enabled.

Author: ```
SELECT
    ROUND(
        (COUNT(*) FILTER (WHERE last_login >= NOW() - INTERVAL '1 year') * 100.0) /
        NULLIF(COUNT(*), 0),
    2) as percentage_active_security_key_users
FROM users
WHERE id IN (SELECT user_id FROM user_security_keys);
```

*   **[NUMBER]** total unique verified logins.

### Trusted Publishing and Attestations

**[Trusted publishing](/posts/2025-11-10-trusted-publishers-coming-to-orgs.md)** continues to be a cornerstone of our security strategy. This year, we've expanded support to include **GitLab Self-Managed instances**, allowing maintainers to automate their release process without needing to manage long-lived API tokens. We've also introduced support for **custom OIDC issuers for organizations**, giving companies more control over their publishing pipelines.

Adoption of trusted publishing has been fantastic:

> *   **[NUMBER]** of projects are now using trusted publishing.

Author: ```
SELECT count(DISTINCT project_id)
FROM oidc_publisher_project_association;
```

Author: ```suggestion
*   more than **50,000** projects are now using trusted publishing
```

> *   **[PERCENTAGE]%** of all uploads to PyPI in the last year were done via trusted publishers.

Author: ```suggestion
*   more than **20%** of all file uploads to PyPI in the last year were done via trusted publishers
```

We've also been hard at work on **attestations**, a new security feature that allows publishers to make verifiable claims about their software. We've added support for attestations from all Trusted Publishing providers, and we're excited to see how the community uses this feature to improve the security of the software supply chain.

> *   **[PERCENTAGE]%** of all uploads to PyPI in the last year that included an attestation.

Author: ```suggestion
*   **17%** of all uploads to PyPI in the last year that included an attestation.
```

## Transparency and Incident Response

This year, we've also focused on providing transparent and timely information about security incidents affecting the PyPI ecosystem. We've published detailed incident reports on a number of events, including:

*   An issue with [privileges persisting in Organization Teams](/posts/2025-04-14-incident-report-organization-team-privileges.md).
*   A widespread [phishing attack targeting PyPI users](/posts/2025-07-31-incident-report-phishing-attack.md).
*   A [token exfiltration campaign via GitHub Actions workflows](/posts/2025-09-16-github-actions-token-exfiltration.md).
*   The potential implications of the ["Shai-Hulud" attack on the npm ecosystem](/posts/2025-11-26-pypi-and-shai-hulud.md).

We believe that transparency is key to building and maintaining trust with our community, and we'll continue to provide these reports as needed.

## Introducing Organizations

One of our biggest announcements this year was the general availability of **Organizations** on PyPI. Organizations provide a way for companies and community projects to manage their packages, teams, and billing in a centralized location.

The response has been overwhelming:

> *   **[NUMBER]** of organizations have been created on PyPI.

Author: ```suggestion
*   **7,742** of organizations have been created on PyPI
```

---
title: "PyPI in 2025: A Year in Review"
description: "A look back at the major changes to PyPI in 2025."
authors:
  - di
date: 2025-12-31
tags:
- new features
- security
- organizations
---

As 2025 comes to a close, it's time to look back at another busy year for the Python Package Index. This year, we've focused on delivering critical security enhancements, rolling out powerful new features for organizations, improving the overall user experience for the millions of developers who rely on PyPI every day, and responding to a number of security incidents with transparency.

<!-- more -->

> But first, let's look at some numbers that illustrate the scale of PyPI in 2025:

Author: ```suggestion
But first, let's look at some numbers that illustrate the sheer scale of PyPI in 2025:
```

*   **[NUMBER]** new files published
*   **[NUMBER]** new projects created
*   **[NUMBER]** petabytes of data transferred
*   **[NUMBER]** billions of requests served

These numbers are a testament to the continued growth and vibrancy of the Python community. As always, a huge thanks to our [sponsors](https://pypi.org/sponsors/) who make this kind of scale possible.

Let's dive into some of the key improvements we've made to PyPI this year.

## Security First, Security Always

> Security is our top priority, and in 2025 we've shipped a number of features to make PyPI more secure than ever before.

Author: ```suggestion
Security is our top priority, and in 2025 we've shipped a number of features to make PyPI more secure than ever.
```

Since rolling out these changes, we've seen:

*   **[PERCENTAGE]%** of active users with non-phishable 2FA enabled.
*   **[NUMBER]** total unique verified logins.

### Trusted Publishing and Attestations

**[Trusted publishing](/posts/2025-11-10-trusted-publishers-coming-to-orgs.md)** continues to be a cornerstone of our security strategy. This year, we've expanded support to include **GitLab Self-Managed instances**, allowing maintainers to automate their release process without needing to manage long-lived API tokens. We've also introduced support for **custom OIDC issuers for organizations**, giving companies more control over their publishing pipelines.

Adoption of trusted publishing has been fantastic:

*   **[NUMBER]** of projects are now using trusted publishing.
*   **[PERCENTAGE]%** of all uploads to PyPI in the last year were done via trusted publishers.

> We've also been hard at work on **attestations**, a new security feature that allows publishers to make verifiable claims about their software. We've added support for attestations from all Trusted Publishing providers, and we're excited to see how the community uses this feature to improve the security of the software supply chain.

Author: ```suggestion
We've also been hard at work on **attestations**, a security feature that allows publishers to make verifiable claims about their software. We've added support for attestations from all Trusted Publishing providers, and we're excited to see how the community uses this feature to improve the security of the software supply chain.
```

### Malware Response

We've continued to improve our malware detection and response capabilities. This year, we've processed **more than 2000 malware reports**. This is a testament to the vigilance of our community and the dedication of our administrators.

Our goal is to reduce the time it takes to remove malware from PyPI, and we're happy to report that we're making significant progress:
in the last year, 66% of all reports were handled within 4 hours, climbing to 92% within 24 hours, with only a few more complex issues reaching the maximum of 4 days to remediate.

### Support Requests

Our support team has also been hard at work making sure our users can continue to be effective on PyPI. This year, we've successfully resolved **2221** individual account recovery requests.

We've also handled more than 500 project name retention sequests (PEP 541). This includes an average first triage time less than 1 week. This is a significant improvement compared to the previous 9-month backlog, and we're happy to report that the backlog is current for the month of December.

## Introducing Organizations

> One of our biggest announcements this year was the general availability of **Organizations** on PyPI. Organizations provide a way for companies and community projects to manage their packages, teams, and billing in a centralized location.

Author: ```suggestion
One of our biggest announcements in previous years was the general availability of **organizations** on PyPI. Organizations provide a way for companies and community projects to manage their packages, teams, and billing in a centralized location.
```
